### PR TITLE
[FIX] payment: update jquery.payment from 1.0.2 to 3.0.0

### PR DESCRIPTION
**Pull request by adr-odoo**
After deep comparison between majors versions, it seems that
there is no breaking changes and this fix a lot of issues in
mobile.

e.g.:
Before this update, the validity date wasn't correctly formated
and the end user had to enter the right format himself.
Some users were blocked at the payment process.

Note that this library is only used for acquirer forms (S2S Form
Template) when "Redirection to the acquirer website" is checked
(payment_flow == 's2s') in the acquirer configuration.

We are aware that this library is deprecated and should be replaced
in master.
See https://github.com/stripe/jquery.payment

opw-1986545

### CODE REVIEW

**Review by nim-odoo — CHANGES_REQUESTED** [review]
IMHO not worth the risk to merge this in stable.

**Comment by Yenthe666** [comment]
I got a little hearth attack seeing this PR going against 11.0. 😆

**Comment by odony** [comment]
I've just discussed it with @adr-odoo. He proposed to upgrade to 3.0.0 directly because there is surprisingly little difference between the 1.0 series and the 3.0 series: the main changes done in the meantime revolved around adding support for Elo cards, and then dropping it, i.e. almost nil. You can view the changelog [here](https://github.com/stripe/jquery.payment/releases).

Now, our original bug happens under android in s2s payment registration mode, and users have reported being unable to enter a working expiry date: indeed they need to manually enter spaces around the separator slash, which is hard to understand (while it works correctly on desktop). The root cause for this issue is the inconsistent list of native events triggered by mobile devices, and the upstream bugfix is stripe/jquery.payment@273d79a08c82795eb24ec36463055e973abfe43c (CS fix) -> stripe/jquery.payment@041b0fe25cac13f37dcc1b3a71497e191cc515cb (updated JS).

If we look at the diff between version 1.0.2 (our current version) and the release just after *that* upstream bugfix (1.1.0), we get [this diff](https://github.com/stripe/jquery.payment/compare/v1.0.2...v1.1.0#diff-82fef9eeefe5e97634417fa43a48d9bd) (note: only the changes to `lib/jquery.payment.js` are relevant). Review it closely, you will notice that the main changes are the actual bugfix we care about, plus a few extra changes to support updated patterns for credit card brands (those come with little risk because the author is Stripe - it's their business to verify credit cards, and the API did not change)

If we consider that:
- we can't fully ignore the bug, and there is little chance of fixing it without backporting stripe/jquery.payment@041b0fe25cac13f37dcc1b3a71497e191cc515cb
- upgrading to 1.1.0 gets us a consistent release with stripe/jquery.payment@041b0fe25cac13f37dcc1b3a71497e191cc515cb plus a few CC pattern improvements
- jquery.payment is a small lib used only in [payment_portal.js](https://github.com/odoo/odoo/blob/11.0/addons/payment/static/src/js/payment_portal.js) to validate credit cards

then it seems that a good compromise could be to upgrade to 1.1.0 in 11.0, and upgrade later to 3.0.0 in master (it includes a few extra fixes and features that aren't strictly needed).

@nim-odoo it's your call, cherry-picking the single upstream fix is also an option, or doing nothing at all and telling customers to upgrade ;-)

**Comment by nim-odoo** [comment]
Before discussing the upgrade of the library, I would like to clarify the issue faced.

> users have reported being unable to enter a working expiry date: indeed they need to manually enter spaces around the separator slash, which is hard to understand (while it works correctly on desktop).

This doesn't seem correct.

If we analyze how we handle the expiry date, we can see that the 3 payment acquirers supporting S2S do not require spaces around the slash:

https://github.com/odoo/odoo/blob/aefbd6da12748f078a197e5e3ae0c1cd68b2e6c5/addons/payment_authorize/models/payment.py#L389

https://github.com/odoo/odoo/blob/aefbd6da12748f078a197e5e3ae0c1cd68b2e6c5/addons/payment_ogone/models/payment.py#L537

https://github.com/odoo/odoo/blob/aefbd6da12748f078a197e5e3ae0c1cd68b2e6c5/addons/payment_stripe/models/payment.py#L240-L241

All these pieces of codes retrieves the 2 first and 2 last characters. Spaces (and even the slash) can be omitted. On Odoo's side, all looks good.

However, the mentioned JS library does require a slash, but not spaces:

https://github.com/stripe/jquery.payment/blob/773884cc610d1ed1890bfcf4419d9b9bf8825a97/lib/jquery.payment.js#L433

It means that a value such as `MM/YY` should be accepted in Odoo v11. This can be verified in a standard v11 (on my phone, Opera Browser): https://drive.google.com/file/d/1VqiQ-JY6vfGRlMNiA-4npr8xqnn1W_G8/view?usp=sharing

The example is with Stripe although I'm convinced it works with Authorize and Ogone.

Therefore, the only issue **reported** is that the slash is not added automatically on mobile, while it is on desktop. In my opinion, this is far from being critical. From my own experience, this is quite common on online shops. Many web shops require the user to explicitly write the slash. Moreover, it is the standard display of the expiry date on a credit card.

The only problem remaining is the updated credit card patterns. AFAIK, this wasn't reported as an issue, but we could still update the library in master to support these patterns.

> then it seems that a good compromise could be to upgrade to 1.1.0 in 11.0

We might miss subsequent bug fixes if we do so.

In conclusion, I don't think the current issue is critical enough to update the library in a stable version.

**Comment by adr-odoo** [comment]
Ok, I will fix this issue in master only.
Big thanks for your review.

**Comment by mreficent** [comment]
This can be closed by https://github.com/odoo/odoo/pull/33376, right?

**Comment by dbo-odoo** [comment]
Hi all,

Mastercard got some new numbers (https://www.mastercard.us/en-us/issuers/get-support/2-series-bin-expansion.html) that are not supported in the old version of this library. This means that entering a valid card number for Mastercard might not be recognized, will mark the form as invalid and will actually prevent the customer from proceeding with payment (test card: 2223000010089800)
screenshot-745469-11-0-b4f794 runbot16 odoo com-2019 12 12-17_41_38


2 possibilities:
- we re-open this PR and merge in 11.0
- we modify the library in place by adding more patterns here: https://github.com/odoo/odoo/blob/11.0/addons/payment/static/lib/jquery.payment/jquery.payment.js#L67 (missing 22 to 27, compare to 13.0 if you want)

Personnally, I'd be for updating the lib but @nim-odoo will have the last word

**Comment by dbo-odoo** [comment]
will close #35139

**Comment by dbo-odoo** [comment]
@nim-odoo quick reminder as discussed just now

**Comment by dbo-odoo** [comment]
Fixed by 329a9b5d8e257b4e716e1863ab6fe05ba8d441fd in 11.0 and dd737b9f4621e3f2adc6b9496a6987d412313ebe in 12.0
